clang/lib/Sema/SemaExpr.cpp

```diff
@@ -255,6 +255,7 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, ArrayRef<SourceLocation> Locs,
 
     diagnoseUnavailableAlignedAllocation(*cast<FunctionDecl>(D), Loc);
   }
+  SYCL().EmitDelayedKernelEntryPointDiagnostics(D->getCanonicalDecl());
```

Comment:
I experimented a little. I wonder if we can simply call `CheckSYCLEntryPointFunctionDecl()` here without having to collect partial diagnostics. Take a look at https://godbolt.org/z/T1Pz5T6qf (and enjoy the implementation divergence while you're there!) The diagnostic issued for use of a deleted function isn't issued until the use occurs, but still includes the instantiation stack; see the code starting at line 271 below.